tensorflow/core/kernels/svd_op_gpu.cu.cc

```diff
@@ -0,0 +1,455 @@
+      V[i] = (Scalar) copysign(1.0, V[i]);
+    }
+  }
+}
+
+// Scalar: The input scalar type (can be complex)
+// SScalar: The output type for the singular value,
+//   same as Scalar if real, or the real version if Scalar is complex
+template <class Scalar, class SScalar>
+class SvdOpGpu : public AsyncOpKernel {
+ public:
+  explicit SvdOpGpu(OpKernelConstruction* context) : AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("compute_uv", &compute_uv_));
+    OP_REQUIRES_OK(context, context->GetAttr("full_matrices", &full_matrices_));
+  }
+
+  void RunSVD(OpKernelContext* context, DoneCallback done, int64 m, int64 n,
+              int64 p, int64 batch_size, Scalar* input_ptr,
+              SScalar* outputS_ptr, Scalar* outputU_ptr, Scalar* outputVT_ptr,
+              int* dev_info_ptr, CudaSolver& solver) {
+
+    // Needed for the n=1 fix, see below, since SVD destroys the input
+    Tensor input_copy;
+    if (compute_uv_ && n==1) {
+      OP_REQUIRES_OK_ASYNC(
+        context,
+        context->allocate_temp(DataTypeToEnum<Scalar>::v(), TensorShape({m}), &input_copy),
+        done);
+    }
+
+    for (int64 i = 0; i < batch_size; ++i) {
+      int lda = m;
+      int ldu = m;
+      int ldvt = n;
+      Scalar* input = input_ptr + i * m * n;
+      SScalar* outputS = outputS_ptr + i * p;
+      Scalar* outputU = NULL;
+      Scalar* outputVT = NULL;
+      signed char jobu = 'N';
+      signed char jobvt = 'N';
+
+      // Save input matrix for n=1 fix
+      if (compute_uv_ && n==1) {
+        const GPUDevice& d = context->eigen_device<GPUDevice>();
+        d.memcpy(input_copy.flat<Scalar>().data(), input,
+               m * sizeof(Scalar));
+      }
+
+      if (compute_uv_) {
+        if (full_matrices_) {
+          outputU = outputU_ptr + i * m * m;
+          outputVT = outputVT_ptr + i * n * n;
+          jobu = 'A';
+          jobvt = 'A';
+        } else {
+          outputU = outputU_ptr + i * m * p;
+          outputVT = outputVT_ptr + i * n * p;
+          jobu = 'S';
+          jobvt = 'S';
+        }
+      }
+
+      OP_REQUIRES_OK_ASYNC(
+          context, solver.Gesvd(jobu, jobvt, m, n, input, lda, outputS, outputU,
+                                ldu, outputVT, ldvt, dev_info_ptr + i),
+          done);
+
+      // This is a bug in cuSolver:
+      // If n is one, then outputVT only contains zeros instead of ones.
+      // Hence, I need to fill outputVT manually
+      // The question is: +1 or -1?
+      // -> Compute U*S and compare sign against M
+      // But because S is zero except for the first entry, the multiplication simplifies a lot
+      // However, what happens if M contains zeros? At these indices, it is impossible
+      // to determine the value of V
+      // -> Compute V for all rows in M to cope for zeros.
+      // 1. V' = sum_i (M_i * U_i,1 * S_i)
+      // 2. V = {1, V'>=0, -1, V'<0}
+      // TODO: what is with complex values?
```

Comment:
For complex singular vectors are only unique up to a complex phase factor (rather than just sign for real). You can see how I normalize them in the unit test here: 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/svd_op_test.py#L73

so you can probably use the same idea of computing the mean (complex) product.